extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java

```diff
@@ -0,0 +1,260 @@
+        }
+
+        if (oidcConfig.token.lifespanGrace.isPresent()) {
+            final int lifespanGrace = oidcConfig.token.lifespanGrace.getAsInt();
+            builder.setAllowedClockSkewInSeconds(lifespanGrace);
+        }
+
+        builder.setRelaxVerificationKeyValidation();
+
+        try {
+            JwtConsumer jwtConsumer = builder.build();
+            jwtConsumer.processToClaims(token);
+        } catch (InvalidJwtException ex) {
+            String detail = "";
+            List<ErrorCodeValidator.Error> details = ex.getErrorDetails();
+            if (!details.isEmpty()) {
+                detail = details.get(0).getErrorMessage();
+            }
+            LOG.debugf("Token verification has failed: %s", detail);
+            return Uni.createFrom().failure(ex);
+        }
+        return Uni.createFrom().item(new TokenVerificationResult(OidcUtils.decodeJwtContent(token), null));
+    }
+
+    public Uni<TokenVerificationResult> refreshJwksAndVerifyJwtToken(String token) {
+        return keyResolver.refresh().onItem().transformToUni(new Function<Void, Uni<? extends TokenVerificationResult>>() {
+
+            @Override
+            public Uni<? extends TokenVerificationResult> apply(Void t) {
+                return verifyJwtToken(token);
+            }
+
+        });
+    }
+
+    public Uni<TokenVerificationResult> introspectToken(String token) {
+        return client.introspectToken(token).onItemOrFailure()
+                .transform(new BiFunction<JsonObject, Throwable, TokenVerificationResult>() {
+
+                    @Override
+                    public TokenVerificationResult apply(JsonObject jsonObject, Throwable t) {
+                        if (t != null) {
+                            throw new AuthenticationFailedException(t);
+                        }
+                        if (!Boolean.TRUE.equals(jsonObject.getBoolean(OidcConstants.INTROSPECTION_TOKEN_ACTIVE))) {
+                            throw new AuthenticationFailedException();
+                        }
+                        Long exp = jsonObject.getLong(OidcConstants.INTROSPECTION_TOKEN_EXP);
+                        if (exp != null) {
+                            final int lifespanGrace = client.getOidcConfig().token.lifespanGrace.isPresent()
+                                    ? client.getOidcConfig().token.lifespanGrace.getAsInt()
+                                    : 0;
+                            if (System.currentTimeMillis() / 1000 > exp + lifespanGrace) {
+                                throw new AuthenticationFailedException();
+                            }
+                        }
+
+                        return new TokenVerificationResult(null, jsonObject);
+                    }
+
+                });
+    }
+
+    public void introspectToken(UniEmitter<? super TokenVerificationResult> emitter, String token) {
+        try {
+            emitter.complete(introspectToken(token).await().indefinitely());
```

Comment:
You should use a timeout.

Reply:
I believe it would be simpler to make `introspectToken` synchronous, and make sure they are always called from a thread that can be blocked.